tests/test_reader.py

```diff
@@ -387,3 +387,58 @@ def test_point():
         # Test with COG + Alpha Band
         assert not reader.point(src_dst, [-104.77519499, 38.95367054])[0]
         assert reader.point(src_dst, [-104.77519499, 38.95367054], masked=False)[0] == 0
+
+
+def test_part_with_buffer():
+    """Make sure buffer works as expected."""
+    bounds = [
+        -6574807.42497772,
+        12210356.646387195,
+        -6261721.357121638,
+        12523442.714243278,
+    ]
+    # Read part at full resolution
+    with rasterio.open(COG) as src_dst:
+        img_no_buffer = reader.part(src_dst, bounds, dst_crs=constants.WEB_MERCATOR_CRS)
+
+    x_size = img_no_buffer.width
+    y_size = img_no_buffer.height
+
+    x_res = (bounds[2] - bounds[0]) / x_size
+    y_res = (bounds[3] - bounds[1]) / y_size
+
+    nx = x_size + 4
+    ny = y_size + 4
+
+    # apply a 2 pixel buffer
+    bounds_with_buffer = (
+        bounds[0] - x_res * 2,
+        bounds[1] - y_res * 2,
+        bounds[2] + x_res * 2,
+        bounds[3] + y_res * 2,
+    )
+    with rasterio.open(COG) as src_dst:
+        img = reader.part(
+            src_dst,
+            bounds_with_buffer,
+            height=ny,
+            width=nx,
+            dst_crs=constants.WEB_MERCATOR_CRS,
+        )
+        assert img.width == nx
+        assert img.height == ny
+
+    with rasterio.open(COG) as src_dst:
+        imgb = reader.part(
+            src_dst, bounds, buffer=2, dst_crs=constants.WEB_MERCATOR_CRS
+        )
+        assert imgb.width == nx
+        assert imgb.height == ny
+
+    assert numpy.array_equal(img.data, imgb.data)
+    assert img.bounds == imgb.bounds
+
+    # No resampling is involved. Because we read the full resolution data
+    # all arrays should be equal
+    numpy.array_equal(img_no_buffer.data, imgb.data[:, 2:-2, 2:-2])
+    numpy.array_equal(img_no_buffer.data, img.data[:, 2:-2, 2:-2])
```

Comment:
here we make sure that reading with buffer (manually created and using buffer=) give the same result as non-buffer read